### PR TITLE
James 1731 Mailbox rename & move should keep emails

### PR DESCRIPTION
**Pull request by rouazana**

### CODE REVIEW

**Comment by aduprat** [comment]
:+1: 
Every tests are working on my computer.

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+package org.apache.james.jmap.cassandra;
+
+import org.junit.runner.RunWith;
+
+import cucumber.api.CucumberOptions;
+import cucumber.api.junit.Cucumber;
+
+@RunWith(Cucumber.class)
+@CucumberOptions(features="classpath:cucumber/SetMailboxes.feature",
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java]: SetMailboxes is not a feature

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodStepdefs.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.james.backends.cassandra.EmbeddedCassandra;
+import org.apache.james.jmap.methods.integration.SetMailboxesMethodStepdefs;
+import org.apache.james.mailbox.elasticsearch.EmbeddedElasticSearch;
+import org.apache.james.modules.CassandraJmapServerModule;
+import org.junit.rules.TemporaryFolder;
+
+import cucumber.api.java.After;
+import cucumber.api.java.Before;
+
+public class CassandraSetMailboxesMethodStepdefs {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodStepdefs.java]: @Rule doesn't work with cucumber ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+    public void init() throws Exception {
+        jmapServer.start();
+        RestAssured.port = jmapServer.getJmapPort();
+        RestAssured.config = newConfig().encoderConfig(encoderConfig().defaultContentCharset(Charsets.UTF_8));
+        RestAssured.enableLoggingOfRequestAndResponseIfValidationFails();
+
+        username = "username@" + USERS_DOMAIN;
+        String password = "password";
+        jmapServer.serverProbe().addDomain(USERS_DOMAIN);
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java]: I should be into a background section, see https://github.com/cucumber/cucumber/wiki/Background

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    public void tearDown() {
+        jmapServer.stop();
+    }
+    
+    @Given("^mailbox \"([^\"]*)\" with (\\d+) messages$")
+    public void mailboxWithMessages(String mailboxName, int messageCount) throws Throwable {
+        jmapServer.serverProbe().createMailbox("#private", username, mailboxName);
+        MailboxPath mailboxPath = new MailboxPath(MailboxConstants.USER_NAMESPACE, username, mailboxName);
+        for (int i = 0; i < messageCount; i++) {
+            String content = "Subject: test" + i + "\r\n\r\n"
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java]: extract to a method ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    
+    public void tearDown() {
+        jmapServer.stop();
+    }
+    
+    @Given("^mailbox \"([^\"]*)\" with (\\d+) messages$")
+    public void mailboxWithMessages(String mailboxName, int messageCount) throws Throwable {
+        jmapServer.serverProbe().createMailbox("#private", username, mailboxName);
+        MailboxPath mailboxPath = new MailboxPath(MailboxConstants.USER_NAMESPACE, username, mailboxName);
+        for (int i = 0; i < messageCount; i++) {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java]: `range(0, messageCount).stream().forEach(i -> appendMessage(i))` ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMailboxes.feature [hunk 6]
```diff
@@ -0,0 +1,13 @@
+Feature: SetMailboxes method
+
+Scenario: Renaming a mailbox should keep messages
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMailboxes.feature]: No indentation ?

**Comment by mbaechler** [comment]
really cool

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java (hunk 1 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java]: Yes, but I had no better idea. What do you suggest?

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodStepdefs.java (hunk 2 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodStepdefs.java]: No, unfortunately.

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java (hunk 1 above)
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java]: MailboxModification ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java (hunk 3 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java]: ok

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java (hunk 5 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java]: ok (+ Throwing)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java (hunk 4 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodStepdefs.java]: ok

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMailboxes.feature (hunk 6 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMailboxes.feature]: Note needed.
What would you like? Space based? Tabs?

Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMailboxes.feature]: A scenario is a subelement of a feature, it's usually indented.
See https://github.com/cucumber/cucumber/wiki/Gherkin

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java (hunk 1 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java]: ok

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMailboxes.feature (hunk 6 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMailboxes.feature]: ok

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/MailboxModification.feature [hunk 7]
```diff
@@ -0,0 +1,20 @@
+Feature: Mailbox modification
+  As a James user
+  I want my mails to be kept when modifying a mailbox
+
+  Background:
+    Given a domain named "domain.tld"
+    And a current user with username "username@domain.tld" and password "secret"
+    And a mailbox "INBOX"
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/MailboxModification.feature]: INBOX is not used (:

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/MailboxModification.feature [hunk 8]
```diff
@@ -0,0 +1,20 @@
+Feature: Mailbox modification
+  As a James user
+  I want my mails to be kept when modifying a mailbox
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/MailboxModification.feature]: you probably also want the mailbox to be modified ?

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+package org.apache.james.jmap.cassandra;
+
+import org.junit.runner.RunWith;
+
+import cucumber.api.CucumberOptions;
+import cucumber.api.junit.Cucumber;
+
+@RunWith(Cucumber.class)
+@CucumberOptions(features="classpath:cucumber/MailboxModification.feature",
+                glue={ "org.apache.james.jmap.methods.integration", "org.apache.james.jmap.cassandra"})
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java]: a lost space here

File: server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/MemorySetMailboxesMethodCucumberTest.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+package org.apache.james.jmap.memory;
+
+import org.junit.runner.RunWith;
+
+import cucumber.api.CucumberOptions;
+import cucumber.api.junit.Cucumber;
+
+@RunWith(Cucumber.class)
+@CucumberOptions(features="classpath:cucumber/MailboxModification.feature",
+                glue={ "org.apache.james.jmap.methods.integration", "org.apache.james.jmap.memory"})
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/MemorySetMailboxesMethodCucumberTest.java]: copy/pasted lost space

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/MailboxModification.feature (hunk 8 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/MailboxModification.feature]: ok

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/MailboxModification.feature (hunk 7 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/MailboxModification.feature]: ok

File: server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/MemorySetMailboxesMethodCucumberTest.java (hunk 10 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/MemorySetMailboxesMethodCucumberTest.java]: ok

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java (hunk 9 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraSetMailboxesMethodCucumberTest.java]: ok

**Comment by mbaechler** [comment]
:+1:

**Comment by chibenwa** [comment]
:+1:

**Comment by aduprat** [comment]
:+1:

**Pull request merged**